java-checks/src/main/java/org/sonar/java/checks/synchronization/TwoLocksWaitNotifyCheck.java

```diff
@@ -0,0 +1,182 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.synchronization;
+
+import com.google.common.collect.ImmutableList;
+import org.sonar.check.Rule;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.LambdaExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.SynchronizedStatementTree;
+import org.sonar.plugins.java.api.tree.SyntaxToken;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.sonar.java.model.ModifiersUtils.findModifier;
+import static org.sonar.plugins.java.api.tree.Modifier.SYNCHRONIZED;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.CONSTRUCTOR;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.METHOD;
+
+@Rule(key = "S3046")
+public class TwoLocksWaitNotifyCheck extends IssuableSubscriptionVisitor {
+
+  private static final MethodMatcher WAIT_MATCHER = MethodMatcher.create().name("wait").withoutParameter();
+  private static final MethodMatcher NOTIFY_MATCHER = MethodMatcher.create().name("notify").withoutParameter();
+
+  private Deque<Counter> synchronizedStack = new LinkedList<>();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(METHOD, CONSTRUCTOR);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    if (tree.is(METHOD, CONSTRUCTOR)) {
+      MethodTree methodTree = (MethodTree) tree;
+      int initialCounter = findModifier(methodTree.modifiers(), SYNCHRONIZED).map(m -> 1).orElse(0);
+      synchronizedStack.push(new Counter(initialCounter));
+      methodTree(methodTree);
+    }
+  }
+
+  @Override
+  public void leaveNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    if (tree.is(METHOD, CONSTRUCTOR)) {
+      synchronizedStack.pop();
+    }
+  }
+
+  private void methodTree(MethodTree tree) {
```

Comment:
this a rather generic name: I have no idea reading the name of the method what this does... maybe something : findWaitOrNotifyInvocation ?